### PR TITLE
.dbtignore

### LINKED ISSUE #5733: [CT-1102] [Feature] .dbtignore
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

A `.dbtignore` file that uses `.gitignore` syntax (including pattern matching) to define files that should NOT be parsed/loaded by dbt.

Code for this would go in [`read_files`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/parser/read_files.py)

After thinking about this for 5 minutes, I don't think there would be unintended consequences for partial parsing. If a file is ignored, it's as if it were deleted; if it's un-ignored, it's as if it were created for the first time.

### Describe alternatives you've considered

Regex support for `model-paths`, `test-paths`, etc. Gross

### Who will this benefit?

https://getdbt.slack.com/archives/C03QUA7DWCW/p1661873024869149

Concretely, people who have `.py` files in their dbt projects that are not actually Python models!

### Are you interested in contributing this feature?

_No response_

### Anything else?

Another use case where this has come up: Removing unneeded files in dbt packages (https://github.com/dbt-labs/dbt-core/issues/4837). Unfortunately, because of how dbt installs Hub packages, this would still require downloading all files and then deleting the "ignored" ones. See conversation in https://github.com/dbt-labs/hubcap/issues/95.

### PR DESCRIPTION
**Pull request by ChenyuLInx**
resolves #5733

### Description
Add .dbtignore function

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Review by jtcohen6 — COMMENTED** [review]
Looks much simpler than I was afraid it would be. Big thanks to the maintainer of `pathspec`!

In some local testing, this seems to work just fine with partial parsing. Only real risk here is degraded parsing performance—adding more overhead to `read_files_elapsed`—which I haven't observed to a noticeable degree.

We'll want to document this, I just opened an issue so we don't lose track: https://github.com/dbt-labs/docs.getdbt.com/issues/2043

File: core/dbt/parser/read_files.py
Comment (ChenyuLInx) [reply core/dbt/parser/read_files.py:2]: Thanks!!

File: core/dbt/parser/read_files.py
Comment (ChenyuLInx) [inline core/dbt/parser/read_files.py:171]: Don't really like the pass through here but we either do this or need to refactor the whole read files part of code I think

**Review by gshank — APPROVED** [review]
The signatures for the 'read_files' calls are getting kind of long, but I suspect that we'll be doing some major refactoring in that area sometime in the next year, so I think we can leave it until then to deal with it.

File: core/dbt/clients/system.py
Comment (iknox-fa) [inline core/dbt/clients/system.py:66]: We should really avoid using `os.walk` or `os.path` in favor of `Pathlib`.  All the os.path stuff is old AF and has weird operating system specific gotchas.

Comment (ChenyuLInx) [reply core/dbt/clients/system.py:66]: Agree! I think we probably leave it to the larger refactor? Same as the signatures for `read_files` calls are getting long. I also don't like that but feels like right now might be okay to leave it.

**Pull request merged**
